### PR TITLE
Slight refactoring

### PR DESCRIPTION
**Pull request by kitlangton**
I changed the `Main` object to be slightly more "idiomatic". While there's nothing wrong with building the layer and then using it, it's more common to call `provide` on some effect.

In addition, `ZLayer.scoped` fixes the issue of scope propagation, and will acquire/release the resources with the layer's own lifecycle.

Thanks for the blog, by the way!

### CODE REVIEW

File: core/src/main/scala/Main.scala [hunk 1]
```diff
@@ -1,21 +1,20 @@
 import zio.{Console, Scope, ZIO, ZIOAppDefault, ZLayer}
+import izumi.reflect._
```
Comment (hmemcpy) [inline core/src/main/scala/Main.scala:2]: Is this needed?

**Comment by loicdescotte** [comment]
Seems easier to read indeed
A blog post update would be great too :)

File: core/src/main/scala/Main.scala [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
-        )
-        .build
-        .map(_.get[CarApi])
-        .flatMap { api =>
-          api.register("Toyota Corolla WE98765").flatMap(Console.printLine(_)) *>
-            api.register("VW Golf WN12345").flatMap(Console.printLine(_)) *>
-            api.register("Tesla").flatMap(Console.printLine(_))
-        }
-    }
+    program.provide(
```
Comment (adamw) [inline core/src/main/scala/Main.scala:13]: So here `.provide` will allocate the layer, use it for the given effect, and then release it - releasing any resources that were acquired during the construction phase? It wasn't quite clear what the lifecycle of the layers is looking at the scaladoc and which methods of using a layer trigger which lifecycle events

File: core/src/main/scala/Main.scala [hunk 3]
```diff
@@ -1,21 +1,19 @@
 import zio.{Console, Scope, ZIO, ZIOAppDefault, ZLayer}
 
 object Main extends ZIOAppDefault:
+  val program: ZIO[CarApi, Nothing, Unit] =
```
Comment (adamw) [inline core/src/main/scala/Main.scala:4]: @kitlangton One problem I see here is that we are starting to use environment for yet another purpose, which I'm not sure I can capture as a clear rule :) Previously env was used for contextual values - not for passing dependencies. I know it's nit-picking, but I wanted the example to demonstrate some clear rules :)

So I'd expect sth like `def program(api: CarApi): ZIO[Any, Nothing, Unit]`. Or maybe we should extend the use-cases of the environment to cover "entry point to the application"? After all, we do need to kick things off somehow.

**Comment by adamw** [comment]
@kitlangton Thanks, I've updated the code to use `ZLayer.scoped`. I kept my constrainted environment usage though, and the layers are composed the way there were more or less. I think it might be confusing if we started to use environments for a different purpose - but I might be wrong of course :) Let me know what you think: https://github.com/softwaremill/zio2-structure/blob/master/core/src/main/scala/Main.scala

Btw. the fact that `ZLayer` behaves as `Managed`/`Resource` surprised me a little. I understood `Scope` that it's the *only* way now to implement resources. But it turns out that there still is a resource/managed-like datatype in ZIO, namely `ZLayer`. That's how I arrived at `ZLayer[Scope, ..., ...]` to represent a layer which requires an open scope to create the values it creates.

**Comment by hmemcpy** [comment]
@adamw Using `ZLayer.make`/`.makeSome` and then calling `build` on it is manual work where `.provide` should be used. The make/makeSome previously made their appearance in Kit's zio-magic where they allowed magic-like gluing together (partial) layers, instead of manually using more awkward `>>>`, `>+>` or `ZLayer.identity/.require` combinators. It all ended up calling the macro to do this for you, giving you a nicer compiler error messages and a more minimal way of passing individual layers as varargs.

In all ZIO examples the idiomatic way is to either pull the "core" application into a `val program = ...` then from the `run` method call `program.provide(....).exitCode`, or do this inside the `val program` by wrapping the for-comprehension in parens and `.provide` it inside, ending up with just `program.exitCode` on the top-level.

I understand that you want to keep a very clear distinction between using the R for dependencies versus services, but your final example is just not idiomatic ZIO.

**Comment by adamw** [comment]
@hmemcpy Thanks for the background and the explanation :) Definitely the way the code is currently structured is not idiomatic ZIO 1.0. But what about 2.0? Haven't thing changed?

I am basing my approach on a twitter conversation, so I might be missing some details of course. I agree that `program.provide(....).exitCode` is nice, but then the "root dependency" would have to become "context", for which environment should be used? Maybe this depends on the definition of "context", which I see that is something that varies at run time, while here we have a singleton service.

Luckily `.makeSome` + `.build` isn't that much manual work (it's just two method calls in the entire app :) )

**Comment by sideeffffect** [comment]
In my experience, you'd typically code
```scala
override def run(args: List[String]): URIO[ZEnv, ExitCode] =
  mylayer.build.useForever.exitCode
```
because in all likelihood you're creating a server that is supposed to run indefinitely. (Example taken from [here](https://github.com/Idealiste-cz/ideal-voting-backend/blob/master/ideal-voting-server/src/main/scala/cz/idealiste/idealvoting/server/Main.scala#L16-L17=).)

**Comment by kitlangton** [comment]
There's also a `launch` method on `ZLayer` that runs a layer to completion which may be worth exploring. You could also create another `App` service with its own `start` method that houses all of the registration logic, and then have `ZIO.serviceWithZIO[App](_.start).provide(...)` as the entry point. 

But yeah, lots of options 😄

**Comment by sideeffffect** [comment]
> lots of options

In all seriousness, @kitlangton , this might be a good feedback to the ZIO development team :wink: What you could do is to document what is _the_ blessed option (or a very small number of them) or even keep just one (or few) of the options and remove the rest.

**Comment by janlisse** [comment]
Hi @adamw ,

nice that you've put this together!
One question regarding `ZIO.scoped` in `Main`: Is that really needed? I think a global top level scope will already be handled
by `ZIOAppDefault`.

**Comment by adamw** [comment]
@janlisse Yes you are right, fixed, thanks :) 

https://github.com/softwaremill/zio2-structure/commit/a99c52be1c0ae6e53595a856725730263d253639

**Pull request closed without merging**